### DISCUSSION TITLE
warning when looping transition

### DISCUSSION THREAD

**Discussion by ImDino** [visgl/deck.gl#6888] [Q&A]
this is the only way I can make an icon layer with an animation loop to work, and it works well but I get the following warnings, and I can't get it to work in any other way to avoid these warnings.

I get this warning as soon as it's rendering: ```layer-manager.js:264 deck: Multiple new layers with same id animationIcon```
and this warning on the second loop: ```layer-manager.js:227 deck: Multiple old layers with same id animationIcon```

I tried doing the right way of destructing the props and assigning them key by key in the renderLayers method itself and concatenating the this.props.id with a seperate string and use this.setNeedsUpdate() instead of animationIconLayer.setNeedsUpdate() but it doesn't do anything.

Here is the working "ugly" edition:

```
class AnimationIconLayer extends CompositeLayer {
  shouldUpdateState({changeFlags}) {
    return changeFlags.dataChanged;
  }

  updateState({props, oldProps, changeFlags}) {}

  renderLayers() {
    return [
      new IconLayer(this.props)
    ];
  }
};

let animationIconLayer = {};

const props = {
  data: [
    {
      coordinates: [18.0686, 59.3293],
    },
  ],
  id: 'animationIcon',
  sizeScale: 40,
  getPosition: d => d.coordinates,
  getColor: () => [0, 0, 0,0],
  getIcon: (d) => ({
    url: svgToDataURL(getCircleSvg('black', `white`)),
    width: 100,
    height: 100,
  }),
  alphaCutoff: 0,
  pickable: false,
  getSize: d => 2.2,
  transitions: {
    getSize: {
      duration: 1000,
      enter: () => [0, 1],
      onEnd: (d) => {
        animationIconLayer.setNeedsUpdate();
      },
    },
    getColor: {
      type: 'interpolation',
      stiffness: 0.09,
      duration: 1000,
      enter: () => [0,0,0,255],
      easing: t => t,
    },
  }
};

animationIconLayer = new AnimationIconLayer(props);
const deckAnimationOverlay = new DeckOverlay();
deckAnimationOverlay.setMap(store.mapState.map);
deckAnimationOverlay.setProps({ layers: [animationIconLayer] });```

**Comment by Pessimistress** [comment]
This warning has nothing to do with transition. Your composite layer renders a sub layer that has the same id as the parent. See https://deck.gl/docs/api-reference/core/composite-layer#getsublayerprops

**Reply by ImDino** [reply]
I know, but then another problem appears. When destructing the props and assigning them key by key in the renderLayers method itself I can concatenate this.props.id with a seperate string, but I also lose access to animationIconLayer.setNeedsUpdate(). I thought it would be the same to use this.setNeedsUpdate() instead but it doesn't restart the animation, It does trigger the renderLayers method after that though, but I don't know why this particular way of handling it would have it not re-rendering fully, it should practically be the same.

**Reply by Pessimistress** [reply]
Calling `setNeedsUpdate()` on the compositeLayer doesn't lead to any actual change unless you pass some different props to the child IconLayer. If you just need to redraw call `setNeedsRedraw` instead.

**Reply by ImDino** [reply]
setNeedsRedraw unfortunately doesn't reset transitions it seems.

**Reply by Pessimistress** [reply]
Maybe I'm not understanding your use case. What effect are you trying to achieve? `setNeedsUpdate` and `setNeedsRedraw` are not intended to be called outside of the layer.

**Reply by ImDino** [reply]
I want to know the proper way of having an animation loop, for a transition to repeat itself once it's over.

My code above managed to do this but with warnings, and there is no documentation for this sort of thing.

Short explanation of the code above:

// i declared this variable
let animationIconLayer = {};

// then set that variable to a new composite layer containing an icon layer with a transition
animationIconLayer = new AnimationIconLayer(props);

// and in my transition prop "on end" I call animationIconLayer.setNeedsUpdate();
`onEnd: (d) => {
    animationIconLayer.setNeedsUpdate();
}`

I noticed that this.setNeedsUpdate() instead didn't do anything btw, which it should.

As mentioned, the warnings were:
- as soon as it starts rendering: layer-manager.js:264 deck: Multiple new layers with same id animationIcon
- on the second iteration (first loop): layer-manager.js:227 deck: Multiple old layers with same id animationIcon

Either way, the warnings above wouldn't be a problem if there was "a proper way" to do this that I could implement.

**Reply by Pessimistress** [reply]
To be honest, I don't know why the code works for you, but I don't recommend it - it's not something we intentionally support. 

I have a Codepen here showing how I would implement an animation loop: https://codepen.io/Pessimistress/pen/jOqVKVB?editors=0010
